File: client/api/clienttag.go
package api

import (
	"github.com/0ceanslim/grain/client/core"
	"github.com/0ceanslim/grain/config"
)

// resolveClientTag decides whether grain stamps its `client` tag on an event it
// builds, and under what name. The configured default (#80/#99) is overridden
// per-build by the user's settings slider — `override` is nil when the request
// carries no preference, in which case the default applies. Foreign `client`
// tags are always stripped regardless (see core.ApplyClientTag).
func resolveClientTag(override *bool) (enabled bool, name string) {
	enabled, name = true, core.DefaultClientTagName
	if cfg := config.GetConfig(); cfg != nil {
		enabled = cfg.Client.ClientTag.Enabled
		if cfg.Client.ClientTag.Name != "" {
			name = cfg.Client.ClientTag.Name
		}
	}
	if override != nil {
		enabled = *override
	}
	return enabled, name
}

File: www/static/js/client-tag.js
// Per-user client-tag preference (#99), persisted client-side and sent with each
// build request to override the server default. Absent = on (the default).
(function () {
  "use strict";
  var KEY = "grain.clientTag"; // "on" | "off"
  window.grainClientTag = {
    // enabled() is the value to send with build requests. Default on.
    enabled: function () {
      return localStorage.getItem(KEY) !== "off";
    },
    set: function (on) {
      localStorage.setItem(KEY, on ? "on" : "off");
    },
  };
})();
